source/extensions/filters/network/kafka/messages/offset_commit.h

```diff
@@ -0,0 +1,292 @@
+public:
+  size_t feed(const char*& buffer, uint64_t& remaining) {
+    size_t consumed = 0;
+    consumed += partition_.feed(buffer, remaining);
+    consumed += offset_.feed(buffer, remaining);
+    consumed += metadata_.feed(buffer, remaining);
+    return consumed;
+  }
+  bool ready() const { return metadata_.ready(); }
+  OffsetCommitPartition get() const { return {partition_.get(), offset_.get(), metadata_.get()}; }
+
+protected:
+  Int32Deserializer partition_;
+  Int64Deserializer offset_;
+  NullableStringDeserializer metadata_;
+};
+
+/**
+ * Deserializes bytes into OffsetCommitPartition (api version 1)
+ */
+class OffsetCommitPartitionV1Buffer : public Deserializer<OffsetCommitPartition> {
+public:
+  size_t feed(const char*& buffer, uint64_t& remaining) {
+    size_t consumed = 0;
+    consumed += partition_.feed(buffer, remaining);
+    consumed += offset_.feed(buffer, remaining);
+    consumed += timestamp_.feed(buffer, remaining);
+    consumed += metadata_.feed(buffer, remaining);
+    return consumed;
+  }
+  bool ready() const { return metadata_.ready(); }
+  OffsetCommitPartition get() const {
+    return {partition_.get(), offset_.get(), timestamp_.get(), metadata_.get()};
+  }
+
+protected:
+  Int32Deserializer partition_;
+  Int64Deserializer offset_;
+  Int64Deserializer timestamp_;
+  NullableStringDeserializer metadata_;
+};
+
+/**
+ * Deserializes array of OffsetCommitPartition-s v0
+ */
+class OffsetCommitPartitionV0ArrayBuffer
+    : public ArrayDeserializer<OffsetCommitPartition, OffsetCommitPartitionV0Buffer> {};
+
+/**
+ * Deserializes array of OffsetCommitPartition-s v1
+ */
+class OffsetCommitPartitionV1ArrayBuffer
+    : public ArrayDeserializer<OffsetCommitPartition, OffsetCommitPartitionV1Buffer> {};
+
+/**
+ * Deserializes bytes into OffsetCommitTopic v0 (which is composed of topic name + array of v0
+ * partitions)
+ */
+class OffsetCommitTopicV0Buffer : public Deserializer<OffsetCommitTopic> {
+public:
+  size_t feed(const char*& buffer, uint64_t& remaining) {
+    size_t consumed = 0;
+    consumed += topic_.feed(buffer, remaining);
+    consumed += partitions_.feed(buffer, remaining);
+    return consumed;
+  }
+  bool ready() const { return partitions_.ready(); }
+  OffsetCommitTopic get() const { return {topic_.get(), partitions_.get()}; }
+
+protected:
+  StringDeserializer topic_;
+  OffsetCommitPartitionV0ArrayBuffer partitions_;
+};
+
+/**
+ * Deserializes bytes into OffsetCommitTopic v1 (which is composed of topic name + array of v1
+ * partitions)
+ */
+class OffsetCommitTopicV1Buffer : public Deserializer<OffsetCommitTopic> {
+public:
+  size_t feed(const char*& buffer, uint64_t& remaining) {
+    size_t consumed = 0;
+    consumed += topic_.feed(buffer, remaining);
+    consumed += partitions_.feed(buffer, remaining);
+    return consumed;
+  }
+  bool ready() const { return partitions_.ready(); }
+  OffsetCommitTopic get() const { return {topic_.get(), partitions_.get()}; }
```

Comment:
I think it starts to be pretty obvious here that `OffsetCommitPartitionV0Buffer`, `OffsetCommitPartitionV1Buffer`, `OffsetCommitTopicV0Buffer`, `OffsetCommitTopicV1Buffer` & co. are very similar to each other, being composed of 2 or 3 or 4 sub-deserializers, that then take the deserialization result, and just bundle them up into resulting `T` from `Deserializer<T>` via `{ .... }` constructor

I think we can attack it in a few different ways:
a) keep doing the same stuff, and just cope with the code bloat
b) introduce a new class `CompositeDeserializer` that would take a vector (tuple?) of sub-deserializers, something like

```
class CompositeDeserializer : public Deserializer<T> {
private:
  vector<Deserializer> delegates_; // I would be storing Deserializer<int8> next to Deserializer<string> here, so I might need a pointer I guess
public:
  bool ready() { return delegates_.end()->ready() } // same thing, ready when last deserializer is full
  size_t feed(const char*& buffer, uint64_t& remaining) {
    for (Deserializer &d : delegates) {
      d.feed(data, remaining)
    }
  }

  T get() {
    vector<void*> results = // get result from each of delegates and put them in vector
    return T{results} // needs a constructor that would downcast each of elements from void* to proper type
   }
}
```
What I don't like here I have `void*` flowing around requiring downcasting.

*c)* 
Common super classes looking like "deserializer composed of 2 deserializers"
```
class CompositeDeserializerComposedOf2Deserializers<T, Delegate1T, Delegate2T> {
private:
  Delegate1T d1;
  Delegate2T d2;
public:
  bool ready() { return d2.ready() }
  size_t feed (....) {
     d1.feed(....)
     d2.feed(....)
  }
  T get() {
    return { d1.get(), d2.get() };
  }
}
```
drawbacks : it's the template solution back again ;)

d) bonus consideration: make `OffsetCommitRequest` abstract and add `OffsetCommitRequestV0`, `OffsetCommitRequestV1` concrete classes - it could alleviate some of problems present in b) - as the constructor would always know that field 0 is e.g. topic name; instead of needing to check with `api_version` first

Reply:
Hmm yeah, I think I'm starting to see the issue here, agree that b) is not great. I'm not completely opposed to templates, if you think it's the best way. I would pick what you think is best for now (with an eye towards trying to avoid complex templates if possible) and just make sure all of the header files are really well commented. Then I can take another pass? Sound good?

Reply:
Sure! Thank you - I think it's visible the most when we actually have `N` semi-similar classes, then the benefit of (templated) superclass becomes appearent.